enumerable.rb

```diff
@@ -0,0 +1,202 @@
+module Enumerable
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/PerceivedComplexity
+  # rubocop:disable Lint/ToEnumArguments
+  ## method to check incoming data type
+  def check_data_type(data_type)
+    arr = data_type if data_type.instance_of?(Array)
+    arr = to_a if data_type.instance_of?(Range)
+    arr = data_type if data_type.instance_of?(Hash)
+    arr
+  end
+
+  ## my_each method
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    for i in check_data_type(self)
+      yield(i)
+    end
+    self
+  end
+
+  ## my_each_index method
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    i = 0
+    until i >= check_data_type(self).length
+      yield(check_data_type(self)[i], i)
+      i += 1
+    end
+    self
+  end
+
+  ## my_select method
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    arr = []
+    my_each { |i| arr << i if yield(i) }
+    arr
+  end
+
+  ## my_any method
+  def my_any?(arg = nil)
+    if block_given?
+      my_each { |i| return true if yield(i) }
+      false
+    elsif arg.nil?
+      my_each { |i| return true if i }
+    elsif !arg.nil? && (arg.instance_of? Class)
+      my_each { |i| return true if [i.class, i.class.superclass].include?(arg) }
+    elsif !arg.nil? && arg.instance_of?(Regexp)
+      my_each { |i| return true if arg.match(i) }
+    else
+      my_each { |i| return true if i == arg }
+    end
+    false
+  end
+
+  ## my_all method
+  def my_all?(arg = nil)
+    if block_given?
+      my_each { |i| return false if yield(i) == false }
+      true
+    elsif arg.nil?
+      my_each { |i| return false if i.nil? || i == false }
+    elsif !arg.nil? && (arg.instance_of? Class)
+      my_each { |i| return false unless [i.class, i.class.superclass].include?(arg) }
+    elsif !arg.nil? && arg.instance_of?(Regexp)
+      my_each { |i| return false unless arg.match(i) }
+    else
+      my_each { |i| return false if i != arg }
+    end
+    true
+  end
+
+  ## my_none method
+  def my_none?(arg = nil)
+    if block_given?
+      my_all?(&Proc.new)
+    else
+      my_all?(arg)
+    end
+  end
```

Comment:
- [ ] It should return true if the block never returns true for all elements.
- [ ] It should return false if the block ever returns true for all elements.
- [ ] It should return false if the block ever returns true for all elements::range.
- [ ] It should not mutate the original array.
- [ ] When no block or argument is given, it should return true only if none of the collection members is true.
- [ ] When a Regex is passed as an argument, it should return true if none of the collection matches the Regex.
- [ ] When a Regex is passed as an argument, it should return false if any of the collection matches the Regex.
- [ ] When a pattern other than Regex or a Class is given, it should return true only if none of the collection matches the pattern.